### PR TITLE
Latency

### PR DESCRIPTION
**Pull request by nav-github01001**
## Description
This PR adds latency functionality to the library

Formula = difference between heartbeat and heartbeat_ack, measured in ms for more precision

## Checklist

- [ ] - All code changes have been tested, if any
- [ ] - The documentation has been updated accordingly for any code changes
- [ ] - This PR fixes an already existing issue
- [x] - This PR adds a new feature (new functions, methods, classes, parameters, etc.)
- [ ] - This PR fixes a bug
- [ ] - This PR is a breaking change (removes/renames functions, parameters etc.)
- [ ] - This PR is NOT a code change (changes Documentation files, README.md, CONTRIBUTING.md etc.)

### CODE REVIEW

File: EpikCord/__init__.py [hunk 1]
```diff
@@ -2473,7 +2494,7 @@ def to_dict(self):
         if hasattr(self, "timestamp"):
             final_product["timestamp"] = self.timestamp
         if hasattr(self, "color"):
-            final_product["color"] = self.color
+            final_product["color"] = self.color.value
```
[the remark this answers is not in the document]
Comment (nav-github01001) [reply EpikCord/__init__.py:2497]: just a little side improvement

File: EpikCord/__init__.py [hunk 2]
```diff
@@ -696,6 +699,11 @@ async def wrapper():
                 await self.heartbeat(True)
 
             elif event["op"] == GatewayOpcode.HEARTBEAT_ACK:
+                heartbeat_ack_time = time_ns()
```
Comment (Sigmanificient) [inline EpikCord/__init__.py:702]: `perf_counter_ns` is more suitable for timers

File: EpikCord/__init__.py [hunk 3]
```diff
@@ -696,6 +699,11 @@ async def wrapper():
...
+                self.discord_latency:int = heartbeat_ack_time - self.heartbeat_time
+                if len(self.latencies) > 10:
+                    self.latencies.pop(0) # pop the first latency
```
Comment (Sigmanificient) [inline EpikCord/__init__.py:705]: Popping the first element require rebuilding the list, thus being `O(n)` time complexity, maybe could be by replacing the item `counter % size`

File: EpikCord/__init__.py [hunk 4]
```diff
@@ -2082,11 +2091,29 @@ def __init__(
         )
 
         self.utils = Utils(self)
-
+        self.latencies = []
         self.user: ClientUser = None
```
Comment (Sigmanificient) [inline EpikCord/__init__.py:2095]: typing issue somehow

File: EpikCord/__init__.py [hunk 5]
```diff
@@ -696,6 +699,11 @@ async def wrapper():
                 await self.heartbeat(True)
 
             elif event["op"] == GatewayOpcode.HEARTBEAT_ACK:
+                heartbeat_ack_time = time_ns()
+                self.discord_latency:int = heartbeat_ack_time - self.heartbeat_time
+                if len(self.latencies) > 10:
```
[the remark this answers is not in the document]
Comment (nav-github01001) [reply EpikCord/__init__.py:704]: No

File: EpikCord/__init__.py [hunk 6]
```diff
@@ -2082,11 +2091,29 @@ def __init__(
...
-
+        self.latencies = []
         self.user: ClientUser = None
         self.application: Optional[ClientApplication] = None
         self.sections: List[Any] = []
 
+    @property
+    def latency(self):
+        latency = self.discord_latency
+        return latency
```
Comment (Sigmanificient) [inline EpikCord/__init__.py:2102]: Sound like an useless prop here?

File: EpikCord/__init__.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ZenoABC) [reply EpikCord/__init__.py:704]: Why not ?

**Review by Sigmanificient — CHANGES_REQUESTED** [review]

File: EpikCord/__init__.py (hunk 2 above)
Comment (nav-github01001) [reply EpikCord/__init__.py:702]: Ok I will commit tmrw

**Review by Sigmanificient — CHANGES_REQUESTED** [review]
Just waiting for the use of `perf_counter_ns`, cycling list could be added later ig

**Review by Sigmanificient — APPROVED** [review]

File: EpikCord/__init__.py [hunk 7]
```diff
@@ -611,6 +613,7 @@ class EventHandler:
     def __init__(self):
         self.events = defaultdict(list)
         self.wait_for_events = defaultdict(list)
+        self.latencies = []
```
Comment (TheUntraceable) [inline EpikCord/__init__.py:616]: Use a deque of maxlen 10

File: EpikCord/__init__.py [hunk 8]
```diff
@@ -2082,11 +2091,25 @@ def __init__(
...
         self.utils = Utils(self)
-
+        self.latencies = []
         self.user: ClientUser = None
         self.application: Optional[ClientApplication] = None
         self.sections: List[Any] = []
 
+    @property
+    def latency(self):
+        latency = self.discord_latency
```
Comment (TheUntraceable) [inline EpikCord/__init__.py:2101]: Just do `return self.discord_latency`??

File: EpikCord/__init__.py [hunk 9]
```diff
@@ -2082,11 +2091,25 @@ def __init__(
...
+        return latency
```
Comment (Sigmanificient) [inline EpikCord/__init__.py:2102]:
```suggestion
        return self.discord_latency
```

**Review by Sigmanificient — CHANGES_REQUESTED** [review]

File: EpikCord/__init__.py (hunk 7 above)
Comment (nav-github01001) [reply EpikCord/__init__.py:616]: Dw I implemented that

**Review by Sigmanificient — APPROVED** [review]

**Review by ZenoABC — APPROVED** [review]

**Comment by nav-github01001** [comment]
@theuntraceable  when merge

File: EpikCord/__init__.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (TheUntraceable) [reply EpikCord/__init__.py:699]: No explanation... just close...

**Comment by TheUntraceable** [comment]
Implement for shard class please

**Comment by TheUntraceable** [comment]
Nevermind, Shard inherits from Client so it should work.

**Comment by nav-github01001** [comment]
Merge then

**Comment by TheUntraceable** [comment]
Answer the unresolved conversation first, just curious

File: EpikCord/__init__.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (nav-github01001) [reply EpikCord/__init__.py:699]: 10 is good avg of the network conditions

**Comment by nav-github01001** [comment]
Dunnit

**Pull request merged**